service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/AwaitCloneControlledAzureStorageContainerResourceFlightStep.java

```diff
@@ -0,0 +1,103 @@
+import bio.terra.stairway.exception.FlightWaitTimedOutException;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.common.utils.FlightUtils;
+import bio.terra.workspace.generated.model.ApiClonedControlledAzureStorageContainer;
+import bio.terra.workspace.generated.model.ApiCreatedControlledAzureStorageContainer;
+import bio.terra.workspace.service.job.JobMapKeys;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.storageContainer.ControlledAzureStorageContainerResource;
+import bio.terra.workspace.service.resource.model.StewardshipType;
+import bio.terra.workspace.service.resource.model.WsmResourceType;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import bio.terra.workspace.service.workspace.model.WsmCloneResourceResult;
+import bio.terra.workspace.service.workspace.model.WsmResourceCloneDetails;
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * Wait for the clone storage container flight to complete and add the result to the appropriate map
+ */
+public class AwaitCloneControlledAzureStorageContainerResourceFlightStep implements Step {
+
+  private final ControlledAzureStorageContainerResource resource;
+  private final String subflightId;
+
+  public AwaitCloneControlledAzureStorageContainerResourceFlightStep(
+      ControlledAzureStorageContainerResource resource, String subflightId) {
+    this.resource = resource;
+    this.subflightId = subflightId;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    // wait for the flight
+    try {
+      FlightState subflightState =
+          context.getStairway().waitForFlight(subflightId, FLIGHT_POLL_SECONDS, FLIGHT_POLL_CYCLES);
+      WsmResourceCloneDetails cloneDetails = new WsmResourceCloneDetails();
+      WsmCloneResourceResult cloneResult =
+          WorkspaceCloneUtils.flightStatusToCloneResult(subflightState.getFlightStatus(), resource);
+      cloneDetails.setResult(cloneResult);
+
+      FlightMap resultMap = FlightUtils.getResultMapRequired(subflightState);
+      var clonedContainer =
+          resultMap.get(
+              JobMapKeys.RESPONSE.getKeyName(), ApiClonedControlledAzureStorageContainer.class);
+      cloneDetails.setStewardshipType(StewardshipType.CONTROLLED);
+      cloneDetails.setResourceType(WsmResourceType.CONTROLLED_AZURE_STORAGE_CONTAINER);
+      cloneDetails.setCloningInstructions(resource.getCloningInstructions());
+      cloneDetails.setSourceResourceId(resource.getResourceId());
+      cloneDetails.setDestinationResourceId(
+          Optional.ofNullable(clonedContainer)
+              .map(ApiClonedControlledAzureStorageContainer::getStorageContainer)
+              .map(ApiCreatedControlledAzureStorageContainer::getResourceId)
```

Comment:
What does this syntax do?

Reply:
It calls the function in the map on the object in the `Optional`. I broke it up into separate lines below in case it's helpful

```
Optional<ApiClonedControlledAzureStorageContainer> apiClonedContainerOption = Optional.ofNullable(clonedContainer)

// call getStorageContainer on the ApiClonedControlledAzureStorageContainer inside of the Optional and return result inside an Optional
Optional<ApiCreatedControlledAzureStorageContainer> apiCreatedContainerOption = apiClonedContainerOption.map(ApiClonedControlledAzureStorageContainer::getStorageContainer)

// call getResourceId on the ApiCreatedControlledAzureStorageContainer inside of the Optional and return result inside an Optional
Optional<UUID> resourceIdOption = apiCreatedContainerOption.map(ApiCreatedControlledAzureStorageContainer::getResourceId)
```